### DISCUSSION TITLE
Automatic parallelization&concurrency using the equations of a 'system'/case study

### DISCUSSION THREAD

**Discussion by balacij** [JacquesCarette/Drasil#3198] [Ideas]
The system of equations and definitions used in the case studies expose what quantities they depend on.

For example, if we have 5 quantity definitions (read from 5 to 1):

1. `e = some other large time-consuming operation, independent of shown variables`
2. `d = some large time-consuming operation, independent of shown variables`
3. `b = g(d, e)`
4. `c = some large monte carlo simulation, independent of shown variables`
5. `a = f(b, c)`

We have information that `a` depends on `b` and `c` being computed values, `b` being dependent on `d` and `e` being computed values, and `c`, `d`, and `e` being independent of the shown variables but defined with long-running equations. If we were to generate a program right now with this, I _think_ we only have opportunity for a 'flat' consecutive calculation scheme, but we can probably use more system resources (e.g., threading) to naively make the generated program faster at runtime. Here for instance, we could make `e`, `d`, and `c` independently calculated on separated threads, then have them merged onto the main thread to calculate `b` and `a` appropriately.

I think there is already some groundwork for this in Drasil, with the manually formed "blocks" of equations used in the examples, but I _think_ this is something that we could calculate automatically from a given system.

In case anyone wants to look into this more, with OO languages, I think the common paradigm is to use "Futures" or "Promises" to do things like this, but there are probably alternative ways. There are also likely some extra things that could be done to the bodies of `e`, `d`, and `c` to make them parallelizable too.

Any case, this might also be an interesting case study in Drasil: incorporating more Algebraic Modelling Language features in Drasil.

(Though, this is really just another example of domain-specific interpretation and what using DSLs to capture 'knowledge' allows us to do.)

**Comment by JacquesCarette** [comment]
The way I would envision doing this would be to 'teach' Drasil (and GOOL) some additional concepts about concurrency. This could then be exposed at the 'recipe' level. There is a lot of fun design that can be done around there.

**Reply by balacij** [reply]
> This could then be exposed at the 'recipe' level.

To confirm, what do you mean by "recipe level"? To me, this would either be a configuration option in the code generator or perhaps even a new, separate, code generator that works under different assumptions. 

> There is a lot of fun design that can be done around there.

I agree! I think (hope, really) that there would be some research in this area too. 

Anycase, this might be one of the nice side-effects/benefits of Drasil (and the Drasil ethos): reducing confounding variables in testing variations of designs of things (focused on software, but absolutely does not need to be software-related, just software _artifact_ related). The reason why I think this is because the field of Algebraic Modelling Languages exist, but I'm not 100% sure.

**Comment by smiths** [comment]
Automatic parallelization and concurrency are great ideas.  I'm all for exploring this.  However, for the current examples parallelization wouldn't be worth the effort.  The problems are far too simple.  There are examples where parallelization is very relevant in research software.  The most obvious spots for parallelization are to link to an external library that has already been parallelized.  For instance, a common bottleneck in research software is solving a large linear system of equations.  A parallel solver can be used to decrease the computation time.

We haven't really explored much with the computational knowledge in Drasil, but there isn't any reason we can't add more of this knowledge in the future.  If we taught Drasil about an algorithm, say Gaussian Elimination, and about concurrency, we might be able to generate a parallel algorithm.  This reminds me of the work that Dr. Anand does.  We'd have to do a lot to catch up to the state of the practice on this front, but it would be interesting to explore.

**Reply by balacij** [reply]
Hmm, that's a problem (the examples being too simple) I wasn't expecting.  I'm glad the idea's good.

I don't know what tricks are used by Prime95 to make it run as fast as it does, but re-building a Prime95-like Mersenne Prime search tool might be another example? However, I'm not sure if it would be more or less difficult than solving large linear systems. 

>  This reminds me of the work that Dr. Anand does.

I forgot about this. "HashedExpression" has a few papers out about how they do things -- it might even be usable as the external solver?
